posthog/models/messaging.py

```diff
@@ -19,15 +19,27 @@ def get_or_create(self, defaults=None, **kwargs):
 
         return super().get_or_create(defaults, **kwargs)
 
+    def filter(self, *args, **kwargs):
+        raw_email = kwargs.pop("raw_email", None)
 
-class MessagingRecord(UUIDModel):
+        if raw_email:
+            kwargs["email_hash"] = get_email_hash(raw_email)
+
+        return super().filter(*args, **kwargs)
+
+    def create(self, *args, **kwargs):
+        raw_email = kwargs.pop("raw_email", None)
 
+        if raw_email:
+            kwargs["email_hash"] = get_email_hash(raw_email)
+
+        return super().create(*args, **kwargs)
+
+
+class MessagingRecord(UUIDModel):
     objects = MessagingRecordManager()
 
     email_hash: models.CharField = models.CharField(max_length=1024)
     campaign_key: models.CharField = models.CharField(max_length=128)
     sent_at: models.DateTimeField = models.DateTimeField(null=True)
     created_at: models.DateTimeField = models.DateTimeField(auto_now_add=True)
-
-    class Meta:
-        unique_together = ("email_hash", "campaign_key")  # can only send campaign once to each email
```

Comment:
How sure are you this won't end up sending a bunch of duplicate emails somewhere else?

Reply:
Fairly, since it requires a new param to the _send_email function to get it to send an additional one. I also wrote a test for just that.

Reply:
In thinking about this more I do think we could run into a race condition where it's called more than once in quick succession and we'd send multiple emails. 

Ben and I have gone back and forth about the best way to implement this. I originally just extended MessagingRecord to store timestamps of subsequent sends on it. Because it's a single row and you can lock it while updating we wouldn't run into a race condition. 

But then it seemed like the implementation would be a lot simpler if we just removed the constraint - store multiple rows, just check the resend_frequency against the most recent row. But that introduces other problems 😓 @benjackwhite any thoughts here?

Reply:
I made some updates including adding a new field with a constraint. This is exactly what constraints are designed for so I don't think we should shy away from them tbh.